src/Framework/Framework/Utils/ReflectionUtils.cs

```diff
@@ -348,19 +358,37 @@ public static bool IsCollection(Type type)
             return type != typeof(string) && IsEnumerable(type) && !IsDictionary(type);
         }
 
-        public static bool IsPrimitiveType(Type type)
+        public static bool IsPrimitiveType(this Type type)
         {
             return PrimitiveTypes.Contains(type)
                 || (IsNullableType(type) && IsPrimitiveType(type.UnwrapNullableType()))
-                || type.IsEnum;
+                || type.IsEnum
+                || CustomPrimitiveTypes.GetOrAdd(type, TryDiscoverCustomPrimitiveType) is {};
         }
 
-        public static bool IsSerializationSupported(this Type type, bool includeNullables)
+        private static CustomPrimitiveTypeRegistration? TryDiscoverCustomPrimitiveType(Type type)
         {
-            if (includeNullables)
-                return IsPrimitiveType(type);
+            var attribute = type.GetCustomAttribute<CustomPrimitiveTypeAttribute>();
+            if (attribute == null)
+            {
+                return null;
+            }
+
+            if (IsCollection(type) || IsDictionary(type))
+            {
+                throw new DotvvmConfigurationException($"The type {type} is marked with {nameof(CustomPrimitiveTypeAttribute)}, but it cannot be used as a custom primitive type. Custom primitive types cannot be collections, dictionaries, and cannot be primitive types already supported by DotVVM.");
+            }
+            if (!PrimitiveTypes.Contains(attribute.ClientSideType.UnwrapNullableType()))
+            {
+                throw new DotvvmConfigurationException($"The type {type} is marked with {nameof(CustomPrimitiveTypeAttribute)} but its {nameof(CustomPrimitiveTypeAttribute.ClientSideType)} is not valid. The client-side representation can only be one of the built-in primitive types (string, numbers, date and time types), or a nullable version of such type.");
+            }
+            if (!typeof(ICustomPrimitiveTypeConverter).IsAssignableFrom(attribute.ConverterType))
+            {
+                throw new DotvvmConfigurationException($"The type {type} is marked with {nameof(CustomPrimitiveTypeAttribute)} but its {nameof(CustomPrimitiveTypeAttribute.ConverterType)} doesn't implement the {nameof(ICustomPrimitiveTypeConverter)} interface!");
+            }
 
-            return PrimitiveTypes.Contains(type);
+            var converter = (ICustomPrimitiveTypeConverter)Activator.CreateInstance(attribute.ConverterType)!;
```

Comment:
We should wrap this into a more descriptive exception. Right now if the converter can not be constructed using default ctor, the exception will be ugly